### DISCUSSION TITLE
How to keep reactivity when passing `props` to a Solid component from a third party that's dynamically imported (as a plugin)?

### DISCUSSION THREAD

**Discussion by AlexErrant** [solidjs/solid#1751] [Q&A] [answered]
I'm building [a webapp](https://github.com/AlexErrant/Pentive/) that allows users to load 3rd party plugins. These plugins have access to the DOM. My project is in Solid, and for now let's assume all plugins are also in Solid.

Loading Solid components as plugins works using `<Dynamic/>`:

```ts
const App: Component = () => {
  const [count, setCount] = createSignal(0)
  const [plugin] = createResource(async () => {
    let encodedPlugin = await toModule<{
      default: VoidComponent<{ i: number }>
    }>(script) // function that converts an arbitrary string to a module
    return encodedPlugin.default
  })
  return (
    <div class={styles.App}>
      <button onClick={() => setCount((x) => x + 1)}>Inc {count()}</button>
      <Dynamic component={plugin()} i={count()} />
    </div>
  )
}
```

However, the component is not updating when the prop changes. The exported component is a simple `VoidComponent<{ i: number }>` without a `createRoot`, so I think it should still be under the same reactive context. The `script` can be found [here](https://github.com/AlexErrant/solid-plugin/blob/main/my-solid-project/src/script.ts), which I'm omitting for brevity. It is the result of building [this](https://github.com/AlexErrant/solid-plugin/blob/main/my-solid-plugin/src/Comp.tsx):

```tsx
import { VoidComponent } from "solid-js"

const Comp: VoidComponent<{ i: number }> = (props) => {
  return <h1>Hello world!!! {props.i} x</h1>
}

export default Comp
```

I have a minimal repo [here](https://github.com/AlexErrant/solid-plugin) and its stackblitz is [here](https://stackblitz.com/~/github.com/AlexErrant/solid-plugin).

**Comment by atk** [comment, marked as answer]
The problem here is that the plugin is using it's own instance of solid, which breaks reactivity.

You need to find a way to provide the same instance of solid to your app and to the plugins. The only way I can think of right now is to use external in the rollup options of vite: https://rollupjs.org/configuration-options/

**Reply by AlexErrant** [reply]
Yep, that works, thanks! Pushed a new commit demoing it working on stackblitz. It only works with `vite preview` due to [this issue](https://github.com/vitejs/vite/issues/6582), but that's not Solid's problem :)

As commentary, I have two branches ([on](https://github.com/AlexErrant/solid-plugin/commit/cd56d1b64600996ffddc7bb82490bf6b9c91c84f), [runsWithOwner](https://github.com/AlexErrant/solid-plugin/commit/39750d4dafa3e85a8e309c2f6a8a29cfd841e830)) where I try to get it working with [`on`](https://www.solidjs.com/docs/latest#on) and `runWithOwner`, but neither work. When I look at the compiled output, both define [`let Effects = null;`](https://github.com/AlexErrant/solid-plugin/blob/cd56d1b64600996ffddc7bb82490bf6b9c91c84f/my-solid-project/src/script.ts#L19), which I think is the reactive context/stack that Ryan's mentioned in various media. Only with `rollupOptions: { external: ["solid-js"]}` does `Effects` disappear, so _both_ my main project and plugin use the externalized solid-js module with its `Effects` state. I kinda thought that on/runWithOwner might give me access to the root Effects, but empirically that didn't pan out.
